### ISSUE TITLE
Github integration

### ISSUE THREAD

**Issue by yaskevich** [jonorthwash/ud-annotatrix#376]
Supershort formulations of tasks:
- add fork from GitHub url button to the /annotatrix page
-  add fork from GitHub url textbox to the / page
- support for different url schemes (e.g. raw.githubusercontent.com, github.com, etc.)
- button to submit a pull request
- show diffs before finalizing the PR

It seems that code to interact with Github is already in the app code base. But there is no info, how user could start to use this feature, also it is not clear from which step development should continue.

**Comment by jonorthwash** [comment]
The first step is to implement loading of a corpus from a GitHub url.

**Comment by yaskevich** [comment]
Ok, I tested, it works.
Maybe it is reasonable to set limit of size for downloadable corpus, since I found some issues related big files from GitHub (#377) .

**Comment by jonorthwash** [comment]
No, we allow whatever the user's system won't crash on, and file a bug against notatrix regarding inefficient parsing / memory usage.

**Comment by jonorthwash** [comment]
> Ok, I tested, it works.

Which part?  Was it already implemented??

**Comment by yaskevich** [comment]
Uploading a corpus from Github link was implemented before, and it works.

As to big files, there is _notatrix_ problem (discussed there #377 and _notatrix_ issue is there: https://github.com/keggsmurph21/notatrix/issues/6).

**Comment by jonorthwash** [comment]
> Uploading a corpus from Github link was implemented before, and it works.

That's great, but the remaining pieces of this issue still need to be implemented.

**Comment by yaskevich** [comment]
I've managed to connect the app to Github. As I can see, there are some code parts that are related to Github, but it's more like a draft. 
When user's credentials are put in the app config and user clicks login button, OAuth authentication is initiated which leads to appearance of two menu items  "Manage repositories" and "Manage permissions" (but they don't provide an interface to manage anything).
Current goal is fork, next is PR. I separated functions of loading content from loading and forking.

![2019-07-04 17_28_42-UD Annotatrix](https://user-images.githubusercontent.com/522902/60689655-6a904b80-9ec8-11e9-9605-5d272c7c5812.png)

As I understand, a full workflow could be like this:

- user clicks "fork" in the app
- modal with URL input field (URL of specific corpus) is opened, user puts a link to a corpus.
- name of the repo is extracted from the URL and request to fork the repo in user's github is sent
- corpus is downloaded into the Annotatrix and user is able to edit
- when user saves the state, all changes are stored in the app
- when user presses PR button
  - fork is downloaded
  - the changed corpus is copied to the fork
  - changes are pushed to fork
  - PR for original repo is prepared

**Comment by yaskevich** [comment]
Subtasks of Github integration related to corpus downloading, editing, pushing back, and preparing PR request (back end). Progress is checked.
- [x] Connect to Github
- [x] Request permissions to access a repo
- [x] Prepare DB table to store data of repo
- [ ] Download a file from a repo via...
  - [x] NodeJS request (to be replaced with API)
  - [X] Content API (less than 1 MB)
  - [ ] Tree API (big files)
- [ ] Store the repo metadata in DB
- [x] Fork a repo
- [ ] Push changes to user's fork
- [ ] Prepare PR request to original repo
- [ ] Remove DB data of the repo, if corpus is deleted

_Note for previous workflow description: it is not necessary to make a local repo to push the changes._

**Comment by yaskevich** [comment]
Previous comment is a tracker for the function set for interaction with Github API, if there is someone who would like to keep track of the development of the project, but didn't read this thread before.

**Comment by yaskevich** [comment]
Some thoughts for consideration on challenges related to interaction with files after some experience with Github API.

### **Filesize and Github limits**

Github and its API are designed for dealing rather with small files (less than 1 MB).

Thus, it's easy to get a small file or push/commit it, it all smoothly goes via [Content API ](https://developer.github.com/v3/repos/contents/). But it becomes more tricky, if it's about the files that are bigger than 1 MB. After I made it work, I had to change this way of interacting with Github to [Tree/Blob API](https://developer.github.com/v3/git/blobs/#create-a-blob): so, it's like when I read a directory, look for a file I need, get its meta, and then fetch it as a blob and decode.

This API allows to process files which are bigger than 1 MB, but it's noticed: "[This API supports blobs up to 100 megabytes in size](https://developer.github.com/v3/git/blobs/)."
Although Github is unhappy when user [pushes something bigger than 50 MB](https://help.github.com/en/articles/working-with-large-files).

I use for testing purposes corpora from [Universal Dependencies git](https://github.com/UniversalDependencies/), and I didn't meet any corpus that is larger than 100 MB. Generally, it seems that it's possible to deal with files which are bigger by means of  [Git Large File Storage](https://github.com/git-lfs/git-lfs/blob/master/docs/spec.md) (I hadn't had yet any experience with that thing). However, it would be an issue for Annotatrix.

As to _updating_ a corpus from the app, the siuation is similar to data loading (easy for small files and tricky for bigger ones). So, pushing changes to Github in code is a sequence of queries to low-level API of Github. Technically, it's like compiling commit and replacing a whole tree, not just single request (but 7 requests).

As to files bigger than 100 MB, the only way could be  a programmatic interaction with LFS, which is rather not documented, being super rare case for using Github via code.

Generally, it looks like the larger file size becomes common for UD users, the worse Github suits as a storage service.
E.g., on [limits for repos](https://help.github.com/en/articles/what-is-my-disk-quota): "We recommend repositories be kept under 1GB each. Repositories have a hard limit of 100GB. If you reach 75GB you'll receive a warning from Git in your terminal when you push. This limit is easy to stay within if large files are kept out of the repository. If your repository exceeds 1GB, you might receive a polite email from GitHub Support requesting that you reduce the size of the repository to bring it back down." 

We have to keep those things in mind.
